### PR TITLE
Add test_uploads_and_deletions test

### PR DESCRIPTION
**Pull request by arpad-m**
Adds a test that is a reproducer for (now fixed) #7296 and also for #7707. Likely also for #7244 but I haven't tried that.

The key ordering bug can be reproduced by switching to `merge_delta_keys` instead of `merge_delta_keys_buffered`, so reverting a big part of #7661, although it only sometimes reproduces (30-50% of cases).

part of https://github.com/neondatabase/neon/issues/7554

### CODE REVIEW

File: test_runner/regress/test_compaction.py [hunk 1]
```diff
@@ -190,3 +191,46 @@ def test_sharding_compaction(
...
+    "compaction_algorithm", [CompactionAlgorithm.LEGACY, CompactionAlgorithm.TIERED]
+)
+def test_uploads_and_deletions(
+    neon_env_builder: NeonEnvBuilder,
+    compaction_algorithm: CompactionAlgorithm,
+):
+    """
+    :param compaction_algorithm: the compaction algorithm to use.
+    """
+    neon_env_builder.num_pageservers = 1
```
Comment (problame) [inline test_runner/regress/test_compaction.py:211]: This is the default, don't think it needs to be stated explicitly

File: test_runner/regress/test_compaction.py [hunk 2]
```diff
@@ -190,3 +191,46 @@ def test_sharding_compaction(
...
+        "compaction_target_size": f"{128 * 1024}",
+        # no PITR horizon, we specify the horizon when we request on-demand GC
+        "pitr_interval": "0s",
+        # disable background compaction and GC. We invoke it manually when we want it to happen.
+        "gc_period": "0s",
+        "compaction_period": "0s",
+        # create image layers eagerly, so that GC can remove some layers
+        "image_creation_threshold": "1",
+        "image_layer_creation_check_threshold": "0",
+        "compaction_algorithm": f'{{"kind": "{compaction_algorithm.value}"}}',
```
Comment (problame) [inline test_runner/regress/test_compaction.py:226]: less quoting headache + make the mess that underlyies this explicit  (the suggestion probably needs an import at the top of this file)

```suggestion
        "compaction_algorithm": json.dumps({"kind": compaction_algorithm.value}),
```

File: test_runner/regress/test_compaction.py [hunk 3]
```diff
@@ -190,3 +191,46 @@ def test_sharding_compaction(
...
+    }
+    env = neon_env_builder.init_start(initial_tenant_conf=tenant_conf)
+
+    # TODO remove these allowed errors
+    # https://github.com/neondatabase/neon/issues/7707
+    env.pageserver.allowed_errors.extend(
+        [".*duplicated L1 layer.*", ".*delta layer created with.*duplicate values.*"]
+    )
+
+    generate_uploads_and_deletions(env, pageserver=env.pageserver)
```
Comment (problame) [inline test_runner/regress/test_compaction.py:236]: Can we

1. run test first
2. then check that log_contains the messages we're allowing
3. then do the allowed_errors.extend()

? That way, we'd assert that this test is indeed a reproducer. Or is it too flaky?

**Review by problame — COMMENTED** [review]
The `def generate_uploads_and_deletions` has been moved and not been modfied otherwise, right?

File: test_runner/regress/test_compaction.py (hunk 2 above)
Comment (arpad-m) [reply test_runner/regress/test_compaction.py:226]: Done

File: test_runner/regress/test_compaction.py (hunk 1 above)
Comment (arpad-m) [reply test_runner/regress/test_compaction.py:211]: Removed

File: test_runner/regress/test_compaction.py (hunk 3 above)
Comment (arpad-m) [reply test_runner/regress/test_compaction.py:236]: the "delta layer created with ... duplicate values" occurs flakily but the "duplicated L1 layer" error occurs reliably I think, so i can invoke `assert_log_contains` for it I think.

Comment (problame) [reply test_runner/regress/test_compaction.py:235]: please do that then, and explain in acomment why we can't assert_log_contains for `duplicated L1 layer`.

(This is basically repeating the PR description. But, it's more likely that people will read the comment than do git blame to check the commit message)

**Review by problame — APPROVED** [review]
Giving advance approval to unblock

File: test_runner/regress/test_compaction.py (hunk 3 above)
Comment (arpad-m) [reply test_runner/regress/test_compaction.py]: added a comment.

File: test_runner/regress/test_compaction.py [hunk 4]
```diff
@@ -190,3 +191,54 @@ def test_sharding_compaction(
...
+            ".*delta layer created with.*duplicate values.*",
+            ".*assertion failed: self.lsn_range.start <= lsn.*",
+            ".*HTTP request handler task panicked: task.*panicked.*",
+        ]
+    )
+
+    generate_uploads_and_deletions(env, pageserver=env.pageserver)
+
+    # We would like to assert for "duplicated L1 layer" here, but there is other errors that
+    # might occur flakily. Therefore, don't assert anything.
```
Comment (problame) [inline test_runner/regress/test_compaction.py:244]: I don't get this, I thought at least one of the `allowed_errors` above reproduces reliably?

Comment (arpad-m) [reply test_runner/regress/test_compaction.py]: Originally I've thought that as well, but it turns out that the "duplicated L1 layer" error does not occur reliably. If the "assertion failed: self.lsn_range.start <= lsn" error occurs, the "duplicated L1 layer" is not reached.

see [this](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7758/9087779519/index.html#suites/f08716c3eb261a35d11a79ed9535ded6/17a1b1c9b4e5e98d) as an example.

I have added some code that iterates over all the allowed errors we are adding and ensures that at least one of them is in the log.

**Review by problame — APPROVED** [review]

**Pull request merged**